.github/workflows/build.yml

```diff
@@ -0,0 +1,88 @@
+name: build
+
+on:
+  # Run on PRs and pushes to the default branch.
+  push:
+    branches: [ master ]
+  pull_request:
+    branches: [ master ]
+  schedule:
+    - cron: "0 0 * * 0"
+
+env:
+  _OS_NAME: linux
+  DART_SITE_ENV_DEFS: 1
+  JEKYLL_ENV: production
+  PUB_ENVIRONMENT: bot.github
+  SITE_LOCALHOST_PORT: 5000
+  SITE_JEKYLL_DESTINATION: _site
+  SITE_JEKYLL_SRC: src
+  TMP: ${{ github.workspace }}/tmp
+  TZ: US/Pacific # normalize build timestamp
+
+jobs:
+  test:
+    runs-on: ubuntu-latest
+    strategy:
+      fail-fast: false
+      matrix:
+        sdk: [dev, beta, stable]
+        task:
+          - ./tool/build.sh --check-links
+          - ./tool/check-code.sh
+          - ./tool/check-code.sh --null-safety
+        experimental: [false]
+        include:
+          - sdk: dev
+            task: ./tool/analyze-and-test-examples.sh
+            experimental: true
+          - sdk: beta
+            task: ./tool/analyze-and-test-examples.sh
+            experimental: true
+          - sdk: stable
+            task: ./tool/analyze-and-test-examples.sh
+            experimental: false
+          - sdk: dev
+            task: ./tool/analyze-and-test-examples.sh --null-safety
+            experimental: true
+          - sdk: beta
+            task: ./tool/analyze-and-test-examples.sh --null-safety
+            experimental: true
+          - sdk: stable
+            task: ./tool/analyze-and-test-examples.sh --null-safety
+            experimental: true
+    steps:
+      - uses: actions/checkout@v2
+        with:
+          # submodules param for checkout action to get site-shared, 'true' might be enough
+          submodules: "recursive"
+      - run: mkdir -p $TMP
+      - uses: actions/setup-node@v1
+        with:
+          node-version: 12
+      - uses: ruby/setup-ruby@v1
+        if: contains(matrix.task, 'build.sh')
+        with:
+          bundler-cache: true
+          ruby-version: 2.6.5
+      - uses: dart-lang/setup-dart@v0.3
+        with:
+          sdk: ${{ matrix.sdk }}
+      - name: Install node dependencies
+        # TOOD: caching requires a lockfile for node deps, so does npm ci.
+        run: npm install
+      - name: Install dart dependencies
+        run: dart pub get
+      - run: ./tool/shared/write-ci-info.sh -v
+        if: contains(matrix.task, 'build.sh')
+      - run: ${{ matrix.task }}
+        continue-on-error: ${{ matrix.experimental }}
+      - run: ./tool/shared/deploy.sh --robots ok default
```

Comment:
I wonder if it would be easier to read if this was a separate job?

Reply:
What I did is closest to what Travis used to do. I'm agnostic, the "easier to read" comes at the cost of overall efficiency and latency because the `build.sh` needs to be run twice.

This is what it would look like as a job instead of a step:
```yaml
  deploy:
    needs: test
    runs-on: ubuntu-latest
    if: |
      github.event == 'push'
      && github.ref == 'refs/heads/master'
      && github.repository == 'dart-lang/site-www'
    steps:
      - uses: actions/checkout@v2
        with:
          # submodules param for checkout action to get site-shared, 'true' might be enough
          submodules: "recursive"
      - run: mkdir -p $TMP
      - uses: actions/setup-node@v1
        with:
          node-version: 12
      - uses: ruby/setup-ruby@v1
        with:
          bundler-cache: true
          ruby-version: 2.6.5
      - uses: dart-lang/setup-dart@v0.3
        with:
          sdk: stable
      - name: Install node dependencies
        # TOOD: caching requires a lockfile for node deps, so does npm ci.
        run: npm install
      - name: Install dart dependencies
        run: dart pub get
      - run: ./tool/shared/write-ci-info.sh -v
      - run: ./tool/build.sh
      - run: ./tool/shared/deploy.sh --robots ok default
        env:
          FIREBASE_TOKEN: ${{ secrets.FIREBASE_TOKEN }}
```

Reply:
Ah, I think what you have originally is good then!

Reply:
I've changed my mind, we should do this because otherwise the deploy step might run before all tests are finished (PR: #2933).